### PR TITLE
[Autoscaler] Demand autoscaler take into account utilized resources

### LINKED ISSUE #10408: [autoscaler] Multi node autoscaler doesn't take into account existing resource usage
### What is the problem?

When we calculate the resources available in the cluster, we should take into account the existing usage from `LoadMetrics`: https://github.com/ray-project/ray/blob/master/python/ray/autoscaler/resource_demand_scheduler.py#L54

Otherwise, we might not scale up even when tasks cannot be scheduled. This should be fixed:
1. Subtract the usage of the node when calculating node resources by looking at LoadMetrics.
2. Add unit tests for this scenario.

### PR DESCRIPTION
**Pull request by wuisawesome**
## Why are these changes needed?

Fixes resource demand autoscaler behavior when gpu's have already been added to the cluster. 

## Related issue number

j
Closes #10408 

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/latest/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested (please justify below)

### CODE REVIEW

File: python/ray/autoscaler/resource_demand_scheduler.py [hunk 1]
```diff
@@ -64,42 +65,51 @@ def get_nodes_to_launch(self, nodes: List[NodeID],
...
 
         unfulfilled = get_bin_pack_residual(node_resources, resource_demands)
         logger.info("Resource demands: {}".format(resource_demands))
         logger.info("Unfulfilled demands: {}".format(unfulfilled))
 
-        nodes = get_nodes_for(self.node_types, node_type_counts,
-                              self.max_workers - len(nodes), unfulfilled)
+        nodes = get_nodes_for(
+            self.node_types, node_type_counts,
+            self.max_workers - len(nodes) - len(pending_nodes), unfulfilled)
```
Comment (wuisawesome) [inline python/ray/autoscaler/resource_demand_scheduler.py:79]: incidental bug fix

File: python/ray/autoscaler/autoscaler.py [hunk 2]
```diff
@@ -205,11 +205,21 @@ def _update(self):
...
                     self.launch_new_node(count, node_type=node_type)
 
+        # TODO (Alex): There's a race condition here where `num_pending` is
+        # derived from `pending_launches` and `len(nodes)` comes from
+        # `self.workers` and we have no synchronization to update them at the
+        # same time. This line ensures that we don't underestimate the number
+        # of workers. This may increase the latency of worker startup by an
+        # update (or so) which is better than a correctness issue around
+        # `max_workers`.
+        nodes = self.workers()
```
Comment (ericl) [inline python/ray/autoscaler/autoscaler.py:221]: Should this be inside the `if self.resource_demand_scheduler` block?

File: python/ray/autoscaler/resource_demand_scheduler.py [hunk 3]
```diff
@@ -64,42 +65,51 @@ def get_nodes_to_launch(self, nodes: List[NodeID],
...
+            self, nodes: List[NodeID], pending_nodes: Dict[NodeID, int],
+            usage_by_ip: Dict[str, ResourceDict]
     ) -> (List[ResourceDict], Dict[NodeType, int]):
         """Returns node resource list and node type counts."""
 
         node_resources = []
         node_type_counts = collections.defaultdict(int)
 
-        def add_node(node_type):
+        def add_node(node_type, resources=None):
```
Comment (ericl) [inline python/ray/autoscaler/resource_demand_scheduler.py:92]:
```suggestion
        def add_node(node_type, existing_resource_usage=None):
```

Comment (ericl) [inline python/ray/autoscaler/resource_demand_scheduler.py:92]: Consider renaming this parameter to `existing_resource_usage`

File: python/ray/autoscaler/resource_demand_scheduler.py (hunk 1 above)
Comment (ericl) [reply python/ray/autoscaler/resource_demand_scheduler.py:79]: Can we add a unit test for this one too?

**Review by ericl — CHANGES_REQUESTED** [review]
Close, just a couple suggestions

File: python/ray/autoscaler/autoscaler.py (hunk 2 above)
Comment (wuisawesome) [reply python/ray/autoscaler/autoscaler.py:221]: I think this is the right place (maybe the whitespacing wasn't clear). This is used for the `num_workers` calculation below, and the race is with the `node_launcher` threads.

Comment (ericl) [reply python/ray/autoscaler/autoscaler.py:221]: Isn't the issue that we don't re-read `self.pending_launches.value` after launching more nodes in the block above? Though it does seem there is a race condition where we might miss some recently enqueued node requests if no threads have picked them up.

I would propose instead adding `nodes = self.workers; num_pending = self.pending_launches.value` to the end of that if block at the end instead, to recompute both in this case.

Comment (wuisawesome) [reply python/ray/autoscaler/autoscaler.py:221]: hmm the race condition i was thinking of was the case in which there is a queue (perhaps from the previous run of `_update`), then we get the current worker nodes, a new node launches, then we get the number of pending nodes (so we lose a node).

i think what you're describing is a completely deterministic bug that can be solved by adding refreshing `nodes` inside the resource demand scheduler block

Comment (ericl) [reply python/ray/autoscaler/autoscaler.py:221]: Yeah so we need to refresh both right? Why not do this:

- remove this line
- add to the block above
```
nodes = self.workers
num_pending = self.pending_launches.value
```

File: python/ray/autoscaler/autoscaler.py [hunk 4]
```diff
@@ -482,8 +492,10 @@ def log_info_string(self, nodes, target):
         tmp += self.load_metrics.info_string()
         tmp += "\n"
         if self.resource_demand_scheduler:
+            print(f"{self.load_metrics.get_resource_utilization()}")
```
Comment (ericl) [inline python/ray/autoscaler/autoscaler.py:495]: Please remove the print here or switch it to logger.

**Review by ericl — CHANGES_REQUESTED** [review]
Rest looks good but can you fix the node refresh as commented?

File: python/ray/autoscaler/autoscaler.py (hunk 2 above)
Comment (wuisawesome) [reply python/ray/autoscaler/autoscaler.py:221]: I'll move this if you want, but I don't think we're talking about the same bug

Comment (ericl) [reply python/ray/autoscaler/autoscaler.py:221]: I think this fixes what you're talking about too. If the bug is something related to the existing code, we can maybe address that in a separate PR.

**Review by ericl — APPROVED** [review]

**Pull request merged**